List View: avoid re-rendering all items on block focus. Enable persistent List View in the widget editor.

gwwar opened the pull request:
Proposed changes in this PR move the querying of isSelected to the ListViewBlock to avoid slowing down the editor on block focus.

The key thing to highlight is **block focus** on this branch is **221.22 ms** vs  **930.92 ms** on trunk with List View open. Changes here don't affect first render time which still sits in the 3-5 second range to open List View with the performance test post. **Using a form of windowing should cut first render down to a more respectable `200ms` see** https://github.com/WordPress/gutenberg/pull/35230 

### Functionality Changes

The widget editor was the only consumer of the ListView dropdown (BlockNavigationDropdown). This required some isSelected state to render a subset of items. From the scrollbars in the popover seen in `trunk` this isn't a well known usage, so I'm proposing that we use a ListView sidebar instead for consistency. If this change is too large, I can pull out this change to a separate PR.

| before block selected | before with no blocks selected | After Sidebar |
|-----|-----|-----|
| <img width="1241" alt="CleanShot 2021-10-18 at 10 50 28@2x" src="https://user-images.githubusercontent.com/1270189/137781908-671cf694-47d6-4894-a985-167deeaf76ac.png"> |  <img width="1241" alt="CleanShot 2021-10-18 at 10 50 22@2x" src="https://user-images.githubusercontent.com/1270189/137782009-e8248faa-1e7b-46da-bf9a-fcaabdadc3de.png"> | <img width="1308" alt="CleanShot 2021-10-18 at 15 19 52@2x" src="https://user-images.githubusercontent.com/1270189/137814383-bda9c3c8-5a8e-4abd-9ab8-9b37985d15cf.png"> |


### Testing Instructions

- No regressions in list view
- Expand/Collapse works in all list view instances
- No drag regressions
- Animations are still seen when using the mover controls
- Widget editor inserter, list view and general sidebar interact well.

### Full Performance Results

Edit: note that performance suite changes were removed in a rebase to avoid slowing down performance tests in trunk before landing. (What I measured with can be seen at https://github.com/WordPress/gutenberg/pull/35571/commits/c6b08e5e639767094f550daa8a831f039720bb23)

To give us solid numbers on the proposed changes, I've also included changes in the performance tests to have list view open during `focus` and `typing` in addition to a new test that toggles the list view open and closed.

```
>> post-editor

┌──────────────────────┬──────────────────────────────────────────┬──────────────┐
│       (index)        │ 6d49143e3d63aed79b54ccc10ddac1918b432f79 │    trunk     │
├──────────────────────┼──────────────────────────────────────────┼──────────────┤
│    serverResponse    │                '239.1 ms'                │ '225.86 ms'  │
│      firstPaint      │                '56.3 ms'                 │  '46.98 ms'  │
│   domContentLoaded   │               '275.36 ms'                │ '276.96 ms'  │
│        loaded        │               '287.34 ms'                │  '286.4 ms'  │
│ firstContentfulPaint │               '5541.68 ms'               │ '5432.78 ms' │
│      firstBlock      │               '6302.24 ms'               │ '6226.54 ms' │
│         type         │                '59.29 ms'                │  '61.05 ms'  │
│       minType        │                '53.18 ms'                │  '53.41 ms'  │
│       maxType        │                '72.09 ms'                │  '75.38 ms'  │
│        focus         │               '185.12 ms'                │ '936.54 ms'  │
│       minFocus       │                 '1.2 ms'                 │   '1.2 ms'   │
│       maxFocus       │               '238.13 ms'                │ '1613.51 ms' │
│     inserterOpen     │                '78.58 ms'                │  '80.01 ms'  │
│   minInserterOpen    │                '54.4 ms'                 │  '53.94 ms'  │
│   maxInserterOpen    │               '258.84 ms'                │ '249.32 ms'  │
│    inserterSearch    │                '92.41 ms'                │  '87.78 ms'  │
│  minInserterSearch   │                '58.03 ms'                │  '54.26 ms'  │
│  maxInserterSearch   │               '291.23 ms'                │ '290.97 ms'  │
│    inserterHover     │                '38.41 ms'                │  '38.62 ms'  │
│   minInserterHover   │                 '35 ms'                  │  '32.74 ms'  │
│   maxInserterHover   │                '53.24 ms'                │  '45.12 ms'  │
│     listViewOpen     │               '3358.94 ms'               │ '3074.26 ms' │
│   minListViewOpen    │               '2976.39 ms'               │ '2843.87 ms' │
│   maxListViewOpen    │               '4778.34 ms'               │ '4198.85 ms' │
└──────────────────────┴──────────────────────────────────────────┴──────────────┘

>> site-editor

┌──────────────────────┬──────────────────────────────────────────┬─────────────┐
│       (index)        │ 6d49143e3d63aed79b54ccc10ddac1918b432f79 │    trunk    │
├──────────────────────┼──────────────────────────────────────────┼─────────────┤
│    serverResponse    │               '115.13 ms'                │ '145.1 ms'  │
│      firstPaint      │               '388.67 ms'                │ '468.23 ms' │
│   domContentLoaded   │               '427.13 ms'                │ '513.47 ms' │
│        loaded        │                '590.1 ms'                │ '666.93 ms' │
│ firstContentfulPaint │               '388.67 ms'                │ '468.23 ms' │
│      firstBlock      │               '6572.97 ms'               │ '7006.6 ms' │
│         type         │                '39.62 ms'                │  '38.5 ms'  │
│       minType        │                '35.08 ms'                │ '34.38 ms'  │
│       maxType        │                '58.86 ms'                │ '59.57 ms'  │
└──────────────────────┴──────────────────────────────────────────┴─────────────┘
```

### Problem

While experimenting with the windowing technique in https://github.com/WordPress/gutenberg/pull/35230 we discovered some unexpected behavior with block focus time. 

Focusing on a block in the main editor pane will re-render **all ListViewBlock items**. Depending on if list view is open or not, we can see some dramatic differences in how long it takes to resolve a focus event.

| blocks | focus time list view closed | focus time list view open |
|-----|-----|-----|
| 50 | 87 ms | 114ms |
| 250 | 100 ms | 311 ms |
| 900 | 180 ms | 1000 ms |

If we profile using react devtools, here we can spot some slow renders on each block focus. Long story short, we query for selectedIds on the parent ListView component. In React if a parent updates all child components by default will update regardless of prop changes (which may or may result in DOM updates). So on focus, all items updated.

What we changed here was moving the selectedId querying to the child ListViewBlock component, which as we can see from the performance results above cleared up the slow render.

| List View Will Re-render on block focus | |
|-----|-----|
| <img width="1123" alt="137548678-7e3a2534-7253-4a7f-adb3-827340b2dd27" src="https://user-images.githubusercontent.com/1270189/137548758-ef670ba2-1a9d-491a-8e25-b9a0d5966519.png"> | <img src="https://user-images.githubusercontent.com/1270189/137550428-99eb689c-4c25-4471-84fc-fd8d11b166f4.png"/> |

### TODOs
- [x] Confirm logic changes around `useListViewClientIds` We never pass in selectedClientIds, so perhaps we can remove logic there.
- [x] ~Do we still need the appender code in List View? I didn't find any active references~ Appender code is okay to remove
- [x] Fix/confirm some logic for last of branch styling, and other misc classes.
- [x] ~Verify any differences between async/sync rendering. I ended up removing it in this PR since we no longer had selectedId in the branch to mark items to explicitly render sync~ Moved the nested async call to ListViewBlock
- [x] Can we update widget editor to use a sidebar for list view instead of a dropdown
- [x] ~Can we delete BlockNavigationDropdown, or is this needed for backwards compatibility?~ I opted to not remove the component in this PR

Folks are welcome to hop into the branch directly if you're aware of any needed functionality changes.

gwwar committed:
List View: add e2e utils for list view open and close

gwwar committed:
remove onClick in branch

gwwar committed:
removed unused terminated levels

gwwar committed:
use callback for toggle expand/collapse

gwwar committed:
update path to be a string to avoid ref changes

gwwar committed:
List View: experimental change to move selection querying to List View Block to avoid re-rendering all items on block focus

gwwar commented:
Initial results look promising for the focus case, manually profiling with the performance test post, focus time is back to ~180ms vs ~900ms

<img width="1241" alt="CleanShot 2021-10-16 at 10 46 17@2x" src="https://user-images.githubusercontent.com/1270189/137597187-528ee41c-f37c-4aa3-8740-0197cdd8a5c8.png">

talldan commented on packages/block-editor/src/components/list-view/use-list-view-client-ids.js:
```diff
@@ -7,82 +7,46 @@ import { useSelect } from '@wordpress/data';
-	selectedClientIds,
-	showOnlyCurrentHierarchy
-) =>
+const useListViewClientIdsTree = ( blocks, showOnlyCurrentHierarchy ) =>
```
What do you think about inlining this code? The hooks seem really simple now (which is great!)

gwwar replied:
@talldan do you mean moving the useSelect call to index.js or simplifying code in use-list-view-client-ids.js?

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -6,149 +6,79 @@ import { map, compact } from 'lodash';
+					</Fragment>
 				);
 			} ) }
-			{ hasAppender && (
-				<ListViewAppender
-					parentBlockClientId={ parentBlockClientId }
-					position={ rowCount }
-					rowCount={ appenderPosition }
-					level={ level }
-					terminatedLevels={ terminatedLevels }
-					path={ [ ...path, appenderPosition ] }
-				/>
-			) }
```
I see it was mentioned in the PR description, was it removed because of performance issues?

The appender isn't mentioned in either of the tracking issues, so probably fine to remove. cc @priethor

priethor replied:
Good call, I don't think adding blocks directly in the list view is a top priority as it opens another whole can of worms, but we should at least have a separate discussion/issue; removing it here should be fine.

gwwar replied:
> was it removed because of performance issues?

Yes, I was avoiding needing branch.js to query isSelected state.  Some of the logic around the appender required `isSelected` state, and I didn't find any used references. 

In general, if we have code that is unused, I'd prefer to remove it since it helps reduce package size and makes it easier to maintain. In the future we can always open a new PR to add the appender back.

> Removing it here should be fine.

✨ Excellent!

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -6,149 +6,79 @@ import { map, compact } from 'lodash';
-		! isTreeRoot &&
-		isClientIdSelected( parentClientId, selectedClientIds );
-	const hasAppender = itemHasAppender( parentBlockClientId );
 	// Add +1 to the rowCount to take the block appender into account.
```
The comment could be removed

gwwar replied:
Tidied in https://github.com/WordPress/gutenberg/pull/35706/commits/489ad1bfbd4edcab21039c1938f979e1d0d3e544

talldan commented on packages/block-editor/src/components/list-view/block.js:
```diff
@@ -82,22 +111,42 @@ export default function ListViewBlock( {
-			isLastOfSelectedBranch,
+		// 'is-last-of-selected-branch':
+		// 	withExperimentalPersistentListViewFeatures &&
+		// 	isLastOfSelectedBranch,
```
Could this be deleted?

gwwar replied:
Yes, I'll investigate if we need these classes (and if we do I'll fix logic here).

gwwar replied:
We can fully remove this one, the border radius isn't even visible unless we add a background.

https://user-images.githubusercontent.com/1270189/137807226-85255b8d-d2a3-4477-8c89-9bfd71850e44.mp4

gwwar replied:
Updated in https://github.com/WordPress/gutenberg/pull/35706/commits/401b409e2e6838e1c7cd1cc77208c641a0400571 👍

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -6,149 +6,79 @@ import { map, compact } from 'lodash';
-import { isClientIdSelected } from './utils';
 import { useListViewContext } from './context';
 
 export default function ListViewBranch( props ) {
```
I like the simplification of this component 🎉

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -6,149 +6,79 @@ import { map, compact } from 'lodash';
 				const isDragged = !! draggedClientIds?.includes( clientId );
 
 				return (
-					<AsyncModeProvider key={ clientId } value={ ! isSelected }>
```
I think the root of ListView still has `AsyncModeProvider`, so I think that would make the whole list async, even the selected block. Not seeing any regressions of https://github.com/WordPress/gutenberg/pull/34519 though, so maybe it's ok ... I don't know why though!

gwwar replied:
Good catch, I'll try and tidy that. I mostly moved back to sync here, since we didn't have isSelected, and I know there are some visual glitches if we try rendering all items async.

gwwar replied:
Moving back to sync adds back some focus slowness. I instead moved the nested AsyncModeProvider call down to ListViewBlock in c344ca1

gwwar commented on packages/block-editor/src/components/list-view/use-list-view-client-ids.js:
```diff
@@ -7,82 +7,46 @@ import { useSelect } from '@wordpress/data';
 			return __unstableGetClientIdsTree();
 		},
-		[ blocks, selectedClientIds, showOnlyCurrentHierarchy ]
+		[ blocks, showOnlyCurrentHierarchy ]
```
So I double-checked behavior in trunk and it looks like ListView is still being used as a dropdown in the widget editor.

Currently if `showOnlyCurrentHierarchy` is set, if an item is selected this will render a subset of the tree.

| block selected | with no blocks selected |
|-----|-----|
| <img width="1241" alt="CleanShot 2021-10-18 at 10 50 28@2x" src="https://user-images.githubusercontent.com/1270189/137781908-671cf694-47d6-4894-a985-167deeaf76ac.png"> |  <img width="1241" alt="CleanShot 2021-10-18 at 10 50 22@2x" src="https://user-images.githubusercontent.com/1270189/137782009-e8248faa-1e7b-46da-bf9a-fcaabdadc3de.png"> |

Can we use a sidebar instead in widgets editor? It feels like the popover has limited utility and that folks aren't very aware of this use case. (Note the scrollbars in trunk). If folks agree, I can create a separate PR or add on to this one.

cc @tellthemachines @draganescu @jasmussen

jasmussen replied:
I'd love unifying on a sidebar, if nothing else then just to avoid drift. But I'll defer to @critterverse who may have thoughts on this.

gwwar replied:
Tried this out in https://github.com/WordPress/gutenberg/pull/35706/commits/1557543a248d8e98b9d63c781fd9ad36774b0e04

https://user-images.githubusercontent.com/1270189/137815372-30eafc54-b15b-4833-8aa3-6057beb0d55b.mp4

critterverse replied:
Nice! This could close https://github.com/WordPress/gutenberg/issues/32311.

talldan replied:
+1 to adding it to the widgets screen!

draganescu replied:
Absolutely yes!

gwwar committed:
List View: simplify useListViewClientIds and remove showOnlyCurrentHierarchy prop

gwwar committed:
List View: move nested AsyncModeProvider call to ListViewBlock

gwwar committed:
List View: make available flags present in JSDoc

gwwar committed:
List View: remove unneeded is-last-of-selected-branch logic

gwwar committed:
List View: tidy comments in branch.js

gwwar committed:
Edit Widgets: use list view sidebar instead of dropdown

gwwar commented:
Thanks for the review @talldan @jasmussen @priethor ! 

Scope here increased since I also tried displaying a ListView sidebar for edit-widgets. It'd be lovely to know if folks agree on the Widget ListView sidebar change, or if we should keep the dropdown.

Note that the sidebar change should also be easy to move that to a different PR.  I'm aware that I also need to fix some E2Es from the widget change (I ran out of time today).

draganescu commented:
> It'd be lovely to know if folks agree on the Widget ListView sidebar change, or if we should keep the dropdown.

I am all for removing the dropdown and replacing it with a sidebar! And apparent;ly so is everyone here https://github.com/WordPress/gutenberg/pull/35706#discussion_r731180272 🎉

gwwar committed:
Edit Widgets: make sure global inserter closes on toggle

gwwar committed:
Edit Widgets: fix E2Es and add simple list view test

gwwar commented:
~This one is ready for another review 👍 Two open questions:~

- ~Is it safe to remove https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-navigation/dropdown.js ? Or do we need to keep this one around since it doesn't have the __experimental prefix?~
- ~Should I drop performance suite changes before landing? I updated it to add a listViewOpen test and updated the typing and focus tests to run with ListView open.~

Edit: since I wasn't sure I didn't include these changes.

ramonjd commented:
I have a [small PR](https://github.com/WordPress/gutenberg/pull/35655) going that replaces `<Button />` list elements with `<a />` elements (semantics and accessibility).

I rebased that PR on top of these changes and took it for a spin to check if anything stood out. 

I couldn't spot any obvious regressions on desktop browsers.

In the Widget and Block Editors:

- selecting block list items (nested and root level)
- dragging and dropping block list items 
- mover control animations

all worked as expected. I also tested on iOS Safari.

![Oct-20-2021 10-26-12](https://user-images.githubusercontent.com/6458278/138004112-e0efebee-03c3-4ca2-87a2-d27f68a32019.gif)

Also +1 for adding the sidebar. Thank you!!

jasmussen commented:
Just as a quick test, this is working well for me, seeing no regressions.

I did also try and stress it with drag/dropping in a list view with 900 blocks, and it completely choked, just like trunk, because this is not your windowing PR. So landing both of those would be just great, and if anyone reading this can help review and land them, we'll be in a vastly better place! 👌

gwwar commented:
Thanks for the reviews! This one is ready for another look. Maybe we can get a ✅ from Widget Editor owners? cc @draganescu @critterverse @talldan I'm pretty confidant in the ListView changes and can fix any issues that show up as follow ups if I missed anything.

Since I wasn't sure, I removed the performance suite changes that ran tests with list view open and kept the BlockNavigationDropdown block-editor component. If it is safe to remove BlockNavigationDropdown we can do that in a follow up.

I also updated the summary with the last round of performance results (https://github.com/WordPress/gutenberg/runs/3945157005)

critterverse commented:
This is such a great improvement to the Widgets screen, thank you for working on it @gwwar!

All looks good to me. One minor note that I think we could remove the ellipses menu for the top-level Widget Areas in the List view since you can't really copy those like a regular block:

![widget-area-ellipses-menu](https://user-images.githubusercontent.com/77359364/138158402-281664c4-e56d-4705-a1e5-6a2f0109b68a.jpg)

gwwar committed:
Edit Widgets: do not show actions in list view for widget areas

gwwar commented:
Good eye @critterverse ! I've updated that in 60b90de 

I don't think we'll run into this use case often, so I added an extra flag to ListView to toggle the behavior. (This avoids the need to query block type in each ListViewBlock).

https://user-images.githubusercontent.com/1270189/138176241-59f816b7-251f-484c-9c82-2aac27fe3672.mp4

talldan commented on packages/block-editor/src/components/block-navigation/dropdown.js:
```diff
@@ -67,7 +67,6 @@ function BlockNavigationDropdown(
 
 					<ListView
 						showNestedBlocks
-						showOnlyCurrentHierarchy
```
I think this is a breaking change for the package, so it'd be great to update the package changelog.

gwwar replied:
Updated in https://github.com/WordPress/gutenberg/pull/35706/commits/4d131a30a7e62210ae795b420bbff7e07b72a260 👍

talldan commented on packages/block-editor/src/components/list-view/style.scss:
```diff
@@ -115,6 +112,11 @@
 			}
 		}
 	}
+	//fix focus styling width when one row has fewer cells
```
Pernickety suggestion:
```suggestion
	// Fix focus styling width when one row has fewer cells.
```

gwwar replied:
Good call, updated in https://github.com/WordPress/gutenberg/pull/35706/commits/4d131a30a7e62210ae795b420bbff7e07b72a260

mkaz reviewed (approved):
This looks good and testing well. 👍 

Looking through all previous feedback and looks like it has all been addressed and no concerns. So I think we're good to go. 🚀

gwwar commented:
Thanks @mkaz I'll try landing this tomorrow if folks didn't have any additional feedback!

gwwar committed:
add related changelogs to block-editor and edit-widgets

gwwar committed:
add detail to the performance change

gwwar commented:
@priethor was kind enough to look over the changelog notes, so this one is all set ✨ Thanks for the reviews folks!

jasmussen commented:
🔥
